New CIDR Provider

moofish32 opened the pull request:
This PR adds a new provider called CIDR. The purpose is to ease the pain in dividing up a network into subnets. The interpolation language provides some support, but this provider adds support to carve up IP space by subnet masks in CIDR notations for both IPv6 and IPv4. Many customers probably do not mind the one time effort of carving up IP space, my employer happens to have many VPCs and we manage non-overlapping IP space. Using just the interpolation cidr functions is possible, but not very elegant. This provider was designed to simplify that process.

The provider adds two data sources:

1. Subnet - simple data source to define sequential subnets from a network and offset. Subnets can be easily linked together by referencing one another.
2. Network - simple data source for teams that want to define the network in one place. This requires some repetition because arrays of maps are not available. The benefit is control over ordering versus the subnet approach.

See the documentation and test cases for examples.

I needed to add some networking helper functions: NextSubnet, PreviousSubnet, NetworkValidate, Increment IP, and Decrement IP. I have them in a helper module initially, but I will open a PR against [go-cidr](https://github.com/apparentlymart/go-cidr) and link in the first comment.

moofish32 commented:
Link to go-cidr PR https://github.com/apparentlymart/go-cidr/pull/3

apparentlymart commented:
Hi @moofish32! Thanks for working on this.

This seems like a useful addition, and it definitely seems more usable than the interpolation functions for complicated, systematic setups.

We are currently in the process of splitting the provider plugins into separate repositories, since we've reached some logistical limits of treating it all as one big release, so I'd like to hold on this until after that is done, at which point we can make it its own repo along with the others. Unfortunately that does mean we will need to rewrite things a little to suit the new repo structure, but that should be pretty mechanical.

In the mean time, one thing I am a little worried about is that the way the network data source is defined, with the ordering implied by the keys, seems like it would be easy to mess up and accidentally renumber higher networks when trying to add a new one. Did you consider making the subnet blocks load into a `TypeList` instead, with the definition order creating the numbering order? That way people are safe as long as they only append to the list.

I'll have a look at your go-cidr change soon! Thanks for that, too.

moofish32 commented:
@apparentlymart -- definitely agree on the ordering. I think in my process of discovering the quirks around my original design of an array of maps I completely forgot this could be a list. I'll change that and update.

kwilczynski commented:
This is a nice idea, but I find the way of referencing resulting values quite cumbersome.

moofish32 commented:
@kwilczynski what would you like to see for a solution? I'm open to making some changes.

apparentlymart commented on builtin/providers/cidr/provider_test.go:
```diff
@@ -0,0 +1,48 @@
+	testAccAwsProvider = aws.Provider().(*schema.Provider)
+	testAccProviders = map[string]terraform.ResourceProvider{
+		"cidr": testAccProvider,
+		"aws":  testAccAwsProvider,
```
If at all possible I'd like to avoid depending on any other providers here -- particularly ones that require accounts and can incur costs -- since it makes it more complex and expensive to run the tests, and more likely to fail due to upstream changes, etc.

Do you think it would be reasonable to cut the AWS-specific parts out of the acceptance tests and just assert directly on the data source results? I know this is not as "deep" a test, but I think this will make it more likely that future contributors to this provider will be able to run the tests, and will also smooth the process of us bringing it into the official release process since we won't need to coordinate with giving it access to the AWS acceptance test environment.

moofish32 replied:
@apparentlymart 
Yeah I'm fighting the dependent provider and can't seem to make it work. There appears to be some vendor-ing naming issues, but I'll just walk away from that problem for now.
Should I refactor here to update go-cidr as well (aka update to latest and drop all those extra tests)? Or should I just make a public repo in my github and get this working then let you guys take from there?

joshshand commented:
Hi guys,

Really like the work going on here, it's almost exactly what I'm looking for - thanks @moofish32 / @apparentlymart. Having completed a similar project before (in Ruby + CloudFormation DSL), I'd like to make some suggestions around how this data source could work better and that it may also satisfy comments above:

1. The ability to order the subnet creation i.e. don't rely on alphabetical sorting. Either add an index/order parameter to each subnet declaration or instantiate them in the order they are read from the HCL file (not sure if that's possible whilst being declarative) i.e. top to bottom (I do note that it looks like there may have been some concern around this).

2. It would be great to be able to define a subnet "tier" with the number of AZs required, instead of a subnet for each AZ. I would guess that 9 times out of 10 people will use the same size subnet across AZs but may not use all available AZs. Example:

```hcl
data "cidr_network" "cake" {
	cidr_block = "${var.cidr_block}"
	subnet {
                name = "dmz"
		mask = 24
		azs = 3 
	}
       subnet {
                name = "elb"
		mask = 24
		azs = 3 
       }
}
```

3. Return each subnet "tier" as a list (not sure if this would actually be possible - a map of lists would be ideal but I think support for this is still lacking from core?). If this was possible then your count could then be the length of the list returned for a key, not number of available AZs. Pseudo example:

Returned map
```
{
	"dmz" = ["cidr1","cidr2","cidr3"]
	"private" = ["cidr3","cidr4","cidr5"]
}

```
Resource creation
```hcl
resource "aws_subnet" "dmz" {
	vpc_id            = "${aws_vpc.main.id}"
	count              = "${length(data.cidr_network.cake_subnet_cidr["dmz"])}"
	cidr_block        = "${element(data.cidr_network.cake.subnet_cidrs["dmz"], count.index)}"
	availability_zone = "${data.aws_availability_zones.available.names[count.index]}"
}
```

Keen to discuss the finer points or potentially collaborate on this. Also is there any rough ETA for this to be merged? Even as it stands would be much better than trying to use the inbuilt subnet calculator. 

Thanks again!

moofish32 commented:
@joshshand - thanks for the comments and interest. Let me see if we can answer a few.

1. This is fixed, but they are read in the order they appear in the HCL file. I didn't have a good understanding on my first pass. Fortunately, I received a little guidance.
2. I started with this design. Two things I discovered towards the end of it were that maps of lists or lists of lists are not currently possible to my knowledge. If @apparentlymart or another knowledgable person can point me to an example, then we can build towards this.
3. Without 2. this item isn't possible to my knowledge (again happy to be corrected). 

I settled into this solution as an improvement. I will hopefully have this pulled out into a repo in the next couple of days. @apparentlymart or another maintainer will have to answer ETA, but this team seems pretty busy right now.

moofish32 commented:
@apparentlymart - https://github.com/moofish32/terraform-provider-cidr

A couple of things feel a bit weird to me:
1. I don't have a way to test my website docs right now, is there something coming to explain this better?
2. I feel like I am missing a PR. I have refactored some code which should be a quick review. Let me know if you want me to do something different here.

joshshand commented:
@moofish32 - appreciate the quick reply! Awesome.

1. Sounds good, I think this is the most logical way to structure and divide the supernet. 

2 & 3. Hopefully @apparentlymart will come back to us on this - I feel like an addressable map of lists is sorely missed in many places and is on the backlog of things to do!?

I have cheated somewhat and used something like this in the past, but I don't think it'll work. The resource count is on the number of keys, not the number of list items for any given key:

```hcl
map = {
    "0" = ["obj0-name", "some-param"]
    "1" = ["obj1-name", "some-other-param"]
}

resource "aws_pseudo" "some_resource" {
    count = "${length(var.map)}"
    someproperty0 = "${element(var.map[count.index], 0)}"
    someproperty1 = "${element(var.map[count.index], 1)}"
}
```

Would love to see alternatives or know when more complex maps are supported. 

Also what's the best way to give this a trial, do you suggest compiling from your master branch?

Thanks

apparentlymart commented:
A few different things here so I'm going to try to catch all of them... sorry if I miss anything.

* Testing docs for a new provider is currently a little tricky. [The `terraform-website` repository](https://github.com/hashicorp/terraform-website) has the root scripts and config for the website, and relies on git submodules to include docs from the different providers. For quick "does this look right?" sort of testing it's probably easiest to clone that repo and just copy your docs directory into `source/docs/providers/cidr` and the layout to `source/layouts/cidr.erb` and see if it works, rather than going to the trouble of setting up the new submodule and symlinks. We can do all the final wiring as part of setting up this provider's repository, release process, etc.
* In principle it's possible to create nested structures of various sorts but we usually suggest keeping things flat because it's often easier to explain and less prone to gotchas resulting from the interpolation language's limitations in generic processing of lists. (For example, we don't currently have a construct for "evaluate this expression for each item in this list", which is often a pain for unpacking complex structures that aren't quite the shape needed for a given use-case.)
* As far as ETA is concerned, we're going to hold off on this until after the 0.10 release process is complete since there's a bunch of work required to get a new provider set up with all of the automated testing, etc and the team is currently focused on completing the 0.10 release and its associated existing provider releases.
* That new repo is looking good at first look. Thanks for doing that! To test it in the mean time it should be possible to `go install` the root of that repository, or alternatively to run `make build` (which does the same thing, behind the scenes). If you put the binary that creates in the same dir as `terraform` itself then Terraform should find it and make it available for use.

Indeed not having a PR for the new repo is a tricky wrinkle in this new workflow. The changes in your latest commit look good to me, though. I'll come back and work with you on getting this into the `terraform-providers` organization once we get 0.10.0 final out. In the mean time, now that this is a separate provider repo you can of course start using it as-is by building its separate binary yourself; the only thing you'd be missing here is the auto-install behavior, which will come from it being deployed to the Hashicorp releases server later.

joshshand commented:
Hey @moofish32 finally got everything compiled, the readme.md needs updating regarding the paths to clone and compile etc.

I only just noticed there was additional documentation to define a data "cidr_subnet" instead of a "cidr_network". Due to the limitations of the returned map for cidr_network I am doing some testing with cidr_subnet instead.

I tried the following:
```hcl
variable	"cidr_block" { default = "10.114.0.0/19" }
variable "azs" { default = 3 }

data "aws_availability_zones" "available" {}

data "cidr_subnet" "vpc-services" {
	cidr_block = "${var.cidr_block}"
	subnet_mask = 24
	subnet_count = "${var.azs}"
}

resource "aws_vpc" "main" {
	cidr_block           = "${var.cidr_block}"
	enable_dns_hostnames = true
	enable_dns_support   = true
}

resource "aws_subnet" "subnets_vpc-services" {
	vpc_id            = "${aws_vpc.main.id}"
	count              = "${var.azs}"
	cidr_block        = "${data.cidr_subnet.vpc-services.subnet_cidrs[count.index]}"
	availability_zone = "${data.aws_availability_zones.available.names[count.index]}"
}
```

Which runs without failure however the following is the resultant set of subnets:
```
aws_subnet.subnets_vpc-services[0] - 10.114.0.0/24
aws_subnet.subnets_vpc-services[1] - 10.114.2.0/24
aws_subnet.subnets_vpc-services[2] - 10.114.1.0/24
```

Note [1] & [2], the third octets are flipped. I would have expected:
[0] => .0.
[1] => .1.
[2] => .2.

Any thoughts on this behaviour?

moofish32 commented:
@joshshand -- thanks for kicking off some real world testing. I was [aware](https://github.com/moofish32/terraform-provider-cidr/blob/master/cidr/datasource_cidr_subnet_test.go#L41) that subnet behaved this way. Let me take another look and see if I can find an alternative. 

For the readme updates, are you proposing updating the readme to reference my personal repo for now? Then when this becomes part of the terraform-providers org update the docs to match the org?

moofish32 commented:
@joshshand -- can you pull this branch and test again https://github.com/moofish32/terraform-provider-cidr/pull/1?

moofish32 commented:
@apparentlymart  - hey where are we on adding new providers now that the 0.10 split is out? Are the improved guides for docs or what can I do to help move this along?

Jo-Con-El commented on builtin/providers/cidr/datasource_cidr_subnet.go:
```diff
@@ -0,0 +1,108 @@
+			},
+			"subnet_mask": &schema.Schema{
+				Type:        schema.TypeInt,
+				Description: "The deisred subnet mask to use for creation",
```
*desired*

Jo-Con-El commented on builtin/providers/cidr/datasource_cidr_subnet.go:
```diff
@@ -0,0 +1,108 @@
+	if sap {
+		_, offsetSubnet, perr := net.ParseCIDR(startAfter.(string))
+		if perr != nil {
+			return fmt.Errorf("start_after %v resulted in pasre error %v\n", startAfter, perr)
```
parse

moofish32 replied:
fixed in https://github.com/moofish32/terraform-provider-cidr/pull/2

Jo-Con-El commented on builtin/providers/cidr/datasource_cidr_network.go:
```diff
@@ -0,0 +1,117 @@
+		Schema: map[string]*schema.Schema{
+			"cidr_block": &schema.Schema{
+				Type:        schema.TypeString,
+				Description: "The CIDR Block for the entire network (aka supernet)",
```
block (in lowercase)

moofish32 replied:
fixed in https://github.com/moofish32/terraform-provider-cidr/pull/2

Jo-Con-El commented on builtin/providers/cidr/datasource_cidr_network.go:
```diff
@@ -0,0 +1,117 @@
+	cBlock := d.Get("cidr_block").(string)
+	_, startNet, err := net.ParseCIDR(cBlock)
+	if err != nil {
+		return fmt.Errorf("Error pasring CIDR %v for cidr_network\n", err)
```
*parsing*

moofish32 replied:
fixed in https://github.com/moofish32/terraform-provider-cidr/pull/2

Jo-Con-El commented on builtin/providers/cidr/datasource_cidr_network_test.go:
```diff
@@ -0,0 +1,341 @@
+		}
+		_, _, fail := calculateSubnets(IPNet, tc.SubnetMasks)
+		if fail == nil {
+			t.Errorf("calculateSubnets shoudl have failed starting at %s\n", tc.CurrentSubnet)
```
*should*

moofish32 replied:
fixed in https://github.com/moofish32/terraform-provider-cidr/pull/2

Jo-Con-El commented on builtin/providers/cidr/datasource_cidr_network_test.go:
```diff
@@ -0,0 +1,341 @@
+		}
+		cByName, _, err := calculateSubnets(IPNet, tc.SubnetMasks)
+		if err != nil {
+			t.Errorf("failed calculateSubnets with %v\n", err)
```
I'd rewrite this as:

calculateSubnets failed with...

moofish32 replied:
fixed in https://github.com/moofish32/terraform-provider-cidr/pull/2

Jo-Con-El commented on builtin/providers/cidr/datasource_cidr_network_test.go:
```diff
@@ -0,0 +1,341 @@
+		}
+		for name, got := range cByName {
+			if got != tc.CIDRByName[name] {
+				t.Errorf("Got %v, Expected %v\n", got, tc.CIDRByName[name])
```
expected (lowercase)

moofish32 replied:
fixed in https://github.com/moofish32/terraform-provider-cidr/pull/2

Jo-Con-El commented on builtin/providers/cidr/datasource_cidr_subnet.go:
```diff
@@ -0,0 +1,108 @@
+		Schema: map[string]*schema.Schema{
+			"cidr_block": &schema.Schema{
+				Type:        schema.TypeString,
+				Description: "The CIDR Block for the entire network (aka supernet)",
```
*block* (lowercase)

moofish32 replied:
fixed in https://github.com/moofish32/terraform-provider-cidr/pull/2

sweigle88 commented:
HI, just wanted to understand where this PR is?

joshshand commented:
@sweigle88 looks like there is a branch conflict that still needs to be resolved. Hopefully @moofish32 can comment on it's status...
It'd be great to have this merged as we are using it in anger but obviously as a custom build from source. 👍

fstuck37 commented:
@moofish32 
Any chance you can correct the conflict? 
I would really appreciate this provider being released as it would help me simplify and improve some of our scripts.

Thanks,
Fred

moofish32 commented:
@fstuck37 and @joshshand 

I followed the guide from @apparentlymart in another thread to submit a new provider. I have heard nothing back from Hashicorp at this time. I have the code in personal repository (https://github.com/moofish32/terraform-provider-cidr). I could clean up the docs a bit more, but I was hoping for a little more of a signal that there was a path to acceptance. In light of that perhaps we should open a feature request to support source fields for providers allowing more flexibility by the operator instead of a centralized system for provider acceptance?

The conflicts here are largely because I opened this PR just before the "Great Provider" split and that has a lot of deleted files. 

@apparentlymart if there is another path I should be on let me know.

-Mike

moofish32 commented:
rebased -- and hashicorp has responded about accepting the provider. I am still in the queue.

fstuck37 commented:
@moofish32 This is great news. Looking forward to having this!!!!

Thanks,
Fred

fstuck37 commented:
@apparentlymart, @joshshand
Does anyone know when we can expect this to be added to Terraform?
Thanks,
Fred

moofish32 commented:
@fstuck37  -- I received an email, but was told it would be a  few weeks that was on May 2. I asked about any faster/alternative routes and even proposed that this was small enough we could consider it something worth bringing into Terraform Core. @cgriggs01 was going to take that feedback back to the team and I haven't heard anything since May 17th. I cleaned up my code a bit here: https://github.com/moofish32/terraform-provider-cidr and added Travis to make sure it didn't break while I waited.

cgriggs01 commented:
Hi @moofish32, 

I apologies for the wait, we are trying to move through the reviews as quickly as possible. The [CIDR](https://github.com/moofish32/terraform-provider-cidr) provider is now second on the list to be reviewed by the Terraform engineering team. With the current pace of provider release, I think that we should be able to release this provider in the next month, as long as we don't run into any major complications. :)

Best,
Chris

fstuck37 commented:
@cgriggs01
Thanks for the update. Looking forward to having this available.
This will allow us to rewrite some complex math (that doesn't always work).
Just so you know I did some testing noted in #16266 as well.

Thanks again,
Fred

fstuck37 commented:
@apparentlymart / @cgriggs01,
Hi all was wondering if there was any update on when this would be released?

Thanks,
Fred

fstuck37 commented:
@cgriggs01 
Hi - was wondering if there was any update on this? I'm sure everyone is working hard on .12 but any chance this will make it in before .12 is release or when it is?
Thanks,
Fred

moofish32 commented:
@cgriggs01 or @apparentlymart  -- any update on the process we'll have to jump through to make miracles happen and get official. We seem to be attracting a few potential customers: https://github.com/moofish32/terraform-provider-cidr/issues/9

apparentlymart commented:
Hi @moofish32!

The blocker here is that everyone is 100% focused on finishing up the v0.12.0 release. Once that is done, we'll work on getting this pulled in to `terraform-providers` and installable by `terraform init`. Sorry for the delay!

cgriggs01 commented:
Hey @moofish32,

Sorry for not getting back to you. We are still planning to migrate and release the provider, but unfortunately it's been pushed back until Terraform 0.12 is released. 

I'll let you know when we begin the process

moofish32 commented:
@cgriggs01 @apparentlymart  -- bump here as I finished the 0.12 upgrade.

teamterraform commented:
Hello, we apologize that this pull request review fell by the wayside. We are actively cleaning up Terraform issues and pull requests, and given Terraform's [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10), we will no longer be accepting pull requests for providers into the core repository. We encourage any potential provider to reach out through the [provider development program](https://www.terraform.io/guides/terraform-provider-development-program.html) to become an official provider. Thank you for your ongoing contributions to our community.
